flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/materializedtable/AlterMaterializedTableAsQueryOperation.java

```diff
@@ -0,0 +1,142 @@
+        ResolvedSchema resolvedQuerySchema = queryOperation.getResolvedSchema();
+        ResolvedSchema oldResolvedSchema = oldResolvedMaterializedTable.getResolvedSchema();
+        List<TableChange> tableChanges =
+                validateAndExtractNewColumns(oldResolvedSchema, resolvedQuerySchema).stream()
+                        .map(TableChange::add)
+                        .collect(Collectors.toList());
+        ResolvedSchema newResolvedSchema =
+                new ResolvedSchema(
+                        resolvedQuerySchema.getColumns(),
+                        oldResolvedSchema.getWatermarkSpecs(),
+                        oldResolvedSchema.getPrimaryKey().orElse(null));
+        Schema newSchema = Schema.newBuilder().fromResolvedSchema(newResolvedSchema).build();
+
+        // update schema and definition query
+        String definitionQuery = queryOperation.asSerializableString();
+        CatalogMaterializedTable catalogMaterializedTable =
+                oldResolvedMaterializedTable.getOrigin().copy(newSchema, definitionQuery);
+
+        ResolvedCatalogMaterializedTable newResolvedMaterializedTable =
+                new ResolvedCatalogMaterializedTable(catalogMaterializedTable, newResolvedSchema);
+        ctx.getCatalogManager()
+                .alterTable(newResolvedMaterializedTable, tableChanges, tableIdentifier, false);
+
+        return TableResultImpl.TABLE_RESULT_OK;
+    }
+
+    @Override
+    public String asSummaryString() {
+        Map<String, Object> params = new LinkedHashMap<>();
+        params.put("identifier", tableIdentifier);
+        return OperationUtils.formatWithChildren(
+                "ALTER MATERIALIZED TABLE",
+                params,
+                Collections.singletonList(queryOperation),
+                Operation::asSummaryString);
+    }
+
+    private List<Column> validateAndExtractNewColumns(
+            ResolvedSchema oldSchema, ResolvedSchema newSchema) {
+        List<Column> newAddedColumns = new ArrayList<>();
+
+        if (oldSchema.getColumns().size() > newSchema.getColumns().size()) {
+            throw new ValidationException(
+                    "Cannot alter table. The new schema has fewer columns than the original schema.");
+        }
+
+        for (int i = 0; i < oldSchema.getColumns().size(); i++) {
+            Column oldColumn = oldSchema.getColumns().get(i);
+            Column newColumn = newSchema.getColumns().get(i);
+            if (!oldColumn.equals(newColumn)) {
+                throw new ValidationException(
+                        String.format(
+                                "Cannot alter table. The schema of existing column %s has changed.",
+                                oldColumn.getName()));
+            }
+        }
+        for (int i = oldSchema.getColumns().size(); i < newSchema.getColumns().size(); i++) {
+            Column newColumn = newSchema.getColumns().get(i);
```

Comment:
If the new added column is duplicated with old column, what behavior is? I think you should a test case to verify it.

Reply:
Based on my testing, if duplicate field names appear in a query, the latter field names will automatically have suffixes added, which seems consistent with the logic of the Calcite framework. For example, if there are two fields named `a`, the second a will automatically be renamed to `a0`.

Reply:
I checked the implementation of CreateView, and it does include validation for field name conflicts. We can reuse this logic. https://github.com/apache/flink/blob/75c26e8c9fd39b957fd1e79323b4fae97b2ce4a6/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlNodeConvertUtils.java#L144

Reply:
Can you help create a jira issue to track it? I think we should forbid the duplicated name case.